File: Framework/L2DPose.cpp
/**
 *
 *  You can modify and use this source freely
 *  only for the development of application related Live2D.
 *
 *  (c) Live2D Inc. All rights reserved.
 */
#include "L2DPose.h"

namespace live2d
{
	namespace framework
	{
		L2DPose::L2DPose() :lastModel(NULL),lastTime(0)
		{
		}
		
		
		
		void L2DPose::updateParam(ALive2DModel* model)
		{
			
			if( model != lastModel )
			{
				
				initParam(model);
			}
			lastModel = model;
			
			l2d_int64  curTime = UtSystem::getTimeMSec();
			float deltaTimeSec = ( (lastTime == 0 ) ? 0 : ( curTime - lastTime )/1000.0f);
			lastTime = curTime;
			
			
			if (deltaTimeSec < 0) deltaTimeSec = 0;
			
			int offset=0;
			for (unsigned int i = 0; i < groupRows.size(); i++)
			{
				int rowCount=groupRows[i];
				normalizePartsOpacityGroup(model,deltaTimeSec,offset,rowCount);
				offset+=rowCount;
			}
			copyOpacityOtherParts(model);
		}
		
		
		
		void L2DPose::initParam(ALive2DModel* model)
		{
			int offset=0;
			for (unsigned int i=0; i<groupRows.size(); i++) {
				int rowCount=groupRows[i];
				for (int j = offset; j < offset+rowCount; j++)
				{
					partsGroups[j].initIndex(model);
					int partsIndex = partsGroups[j].partsIndex ;
					int paramIndex = partsGroups[j].paramIndex ;
					
					if(partsIndex<0)continue;
					
					model->setPartsOpacity(partsIndex , (j==offset ? 1.0f : 0.0f) ) ;
					model->setParamFloat(paramIndex , (j==offset ? 1.0f : 0.0f) ) ;
					
					for (unsigned int k = 0; k < partsGroups[j].link.size(); k++)
					{
						partsGroups[j].link[k].initIndex(model);
					}
				}
				offset+=rowCount;
			}
			
		}
		
		
		
		void L2DPose::normalizePartsOpacityGroup( ALive2DModel* model , float deltaTimeSec ,int offset ,int rowCount)
		{
			int visibleParts = -1 ;
			float visibleOpacity = 1.0f ;
			
			float CLEAR_TIME_SEC = 0.5f ;
			float phi = 0.5f ;
			float maxBackOpacity = 0.15f ;
			
			
			
			for (int i = offset ; i < offset + rowCount; i++ )
			{
				int partsIndex = partsGroups[i].partsIndex;
				int paramIndex = partsGroups[i].paramIndex;
				
				if( model->getParamFloat( paramIndex ) != 0 )
				{
					if( visibleParts >= 0 )
					{
						break ;
					}
					
					visibleParts = i ;
					visibleOpacity = model->getPartsOpacity(partsIndex) ;
					
					
					visibleOpacity += deltaTimeSec / CLEAR_TIME_SEC ;
					if( visibleOpacity > 1 )
					{
						visibleOpacity = 1 ;
					}
				}
			}
			
			if( visibleParts < 0 )
			{
				visibleParts = 0 ;
				visibleOpacity = 1 ;
			}
			
			
			for (int i = offset ; i < offset + rowCount ; i++ )
			{
				int partsIndex = partsGroups[i].partsIndex;
				
				
				if( visibleParts == i )
				{
					model->setPartsOpacity(partsIndex , visibleOpacity ) ;
				}
				
				else
				{
					float opacity = model->getPartsOpacity(partsIndex) ;
					float a1 ;
					if( visibleOpacity < phi )
					{
						a1 = visibleOpacity*(phi-1)/phi + 1 ; 
					}
					else
					{
						a1 = (1-visibleOpacity)*phi/(1-phi) ; 
					}
					
					
					float backOp = (1-a1)*(1-visibleOpacity) ;
					if( backOp > maxBackOpacity )
					{
						a1 = 1 - maxBackOpacity/( 1- visibleOpacity ) ;
					}
					
					if( opacity > a1 )
					{
						opacity = a1 ;
					}
					model->setPartsOpacity(partsIndex , opacity ) ;
				}
			}
		}
		
		
		
		void L2DPose::copyOpacityOtherParts(ALive2DModel* model)
		{
			for (unsigned int i_group = 0; i_group < partsGroups.size(); i_group++)
			{
				L2DPartsParam &partsParam = partsGroups[i_group];
				
				if(partsParam.link.size()==0)continue;
				
				int partsIndex = partsGroups[i_group].partsIndex;
				
				float opacity = model->getPartsOpacity( partsIndex );
				
				for (unsigned int i_link = 0; i_link < partsParam.link.size(); i_link++)
				{
					L2DPartsParam &linkParts = partsParam.link[i_link];
					
					int linkPartsIndex = linkParts.partsIndex;
					
					if(linkPartsIndex < 0)continue;
					model->setPartsOpacity(linkPartsIndex, opacity);
				}
			}
		}
		
		
		/**
		 * JSONファイルから読み込む
		 * 仕様についてはマニュアル参照。JSONスキーマの形式の仕様がある。
		 * @param buf
		 * @return
		 */
		L2DPose* L2DPose::load(const void* buf ,int size)
		{
			L2DPose* ret = new L2DPose();
			
			Json* json = Json::parseFromBytes( (const char*)buf , size ) ;
			
			Value& root = json->getRoot() ;
			
			
			Value& poseListInfo = root["parts_visible"];
			int poseNum = poseListInfo.size();
			
			for (int i_pose = 0; i_pose < poseNum; i_pose++)
			{
				Value& poseInfo = poseListInfo[i_pose];
				
				
				Value& idListInfo = poseInfo["group"];
				int idNum = idListInfo.size();
				int rowCount=0;
				for (int i_group = 0; i_group < idNum; i_group++)
				{
					Value& partsInfo=idListInfo[i_group];
					L2DPartsParam parts;
					LDString paramID=partsInfo["id"].toString();
					parts.partsID=paramID;
					
					
					if(partsInfo["link"].isNull())
					{
						
					}
					else
					{
						Value &linkListInfo = partsInfo["link"];
						int linkNum = linkListInfo.size();
						
						for (int i_link = 0; i_link< linkNum; i_link++)
						{
							L2DPartsParam linkParts;
							LDString linkID=linkListInfo[i_link].toString();
							linkParts.partsID=linkID;
							parts.link.push_back(linkParts);
						}
					}
					ret->partsGroups.push_back(parts);
					rowCount++;
				}
				ret->groupRows.push_back(rowCount);
				
			}
			
			delete json;
			return ret;
		}
	}
}

File: package/live2d/v2/framework/pose/l2d_pose.py
from ..Live2DFramework import Live2DFramework
from .l2d_parts_param import L2DPartsParam
from ...core import Array, UtSystem


class L2DPose:

    def __init__(self):
        self.lastTime = 0
        self.lastModel = None
        self.partsGroups = Array()

    def updateParam(self, model):
        if model != self.lastModel:
            self.initParam(model)

        self.lastModel = model
        cur_time = UtSystem.getUserTimeMSec()
        delta_time_sec = (0 if (self.lastTime == 0) else (cur_time - self.lastTime) / 1000.0)
        self.lastTime = cur_time
        if delta_time_sec < 0:
            delta_time_sec = 0
        for i in range(len(self.partsGroups)):
            self.normalizePartsOpacityGroup(model, self.partsGroups[i], delta_time_sec)
            self.copyOpacityOtherParts(model, self.partsGroups[i])

    def initParam(self, model):
        for i in range(len(self.partsGroups)):
            parts_group = self.partsGroups[i]
            for j in range(len(parts_group)):
                parts_group[j].initIndex(model)
                parts_index = parts_group[j].partsIndex
                param_index = parts_group[j].paramIndex
                if parts_index < 0:
                    continue
                v = (model.getParamFloat(param_index) != 0)
                model.setPartsOpacity(parts_index, (1.0 if v else 0.0))
                model.setParamFloat(param_index, (1.0 if v else 0.0))
                if parts_group[j].link is None:
                    continue
                for k in range(len(parts_group[j].link)):
                    parts_group[j].link[k].initIndex(model)

    @staticmethod
    def normalizePartsOpacityGroup(model, partsGroup, deltaTimeSec):
        visible_parts = -1
        visible_opacity = 1.0
        clear_time_sec = 0.5
        phi = 0.5
        max_back_opacity = 0.15
        for i in range(len(partsGroup)):
            parts_index = partsGroup[i].partsIndex
            param_index = partsGroup[i].paramIndex
            if parts_index < 0:
                continue
            if model.getParamFloat(param_index) != 0:
                if visible_parts >= 0:
                    break

                visible_parts = i
                visible_opacity = model.getPartsOpacity(parts_index)
                visible_opacity += deltaTimeSec / clear_time_sec
                if visible_opacity > 1:
                    visible_opacity = 1

        if visible_parts < 0:
            visible_parts = 0
            visible_opacity = 1

        for i in range(len(partsGroup)):
            parts_index = partsGroup[i].partsIndex
            if parts_index < 0:
                continue
            if visible_parts == i:
                model.setPartsOpacity(parts_index, visible_opacity)
            else:
                opacity = model.getPartsOpacity(parts_index)
                if visible_opacity < phi:
                    a1 = visible_opacity * (phi - 1) / phi + 1
                else:
                    a1 = (1 - visible_opacity) * phi / (1 - phi)

                back_op = (1 - a1) * (1 - visible_opacity)
                if back_op > max_back_opacity:
                    a1 = 1 - max_back_opacity / (1 - visible_opacity)

                opacity = min(opacity, a1)

                model.setPartsOpacity(parts_index, opacity)

    @staticmethod
    def copyOpacityOtherParts(model, partsGroup):
        for i_group in range(len(partsGroup)):
            parts_param = partsGroup[i_group]
            if parts_param.link is None:
                continue
            if parts_param.partsIndex < 0:
                continue
            opacity = model.getPartsOpacity(parts_param.partsIndex)
            for link_parts in parts_param.link:
                if link_parts.partsIndex < 0:
                    continue
                model.setPartsOpacity(link_parts.partsIndex, opacity)

    @staticmethod
    def load(buf):
        ret = L2DPose()
        pm = Live2DFramework.getPlatformManager()
        json = pm.jsonParseFromBytes(buf)
        pose_list_info = json.get("parts_visible")
        pose_num = len(pose_list_info)
        for i_pose in range(pose_num):
            pose_info = pose_list_info[i_pose]
            id_list_info = pose_info.get("group")
            id_num = len(id_list_info)
            parts_group = Array()
            for i_group in range(id_num):
                parts_info = id_list_info[i_group]
                parts = L2DPartsParam(parts_info["id"])
                parts_group.append(parts)
                if parts_info.get("link") is None:
                    continue
                link_list_info = parts_info.get("link")
                link_num = len(link_list_info)
                parts.link = Array()
                for i_link in range(link_num):
                    link_parts = L2DPartsParam(link_list_info[i_link])
                    parts.link.append(link_parts)

            ret.partsGroups.append(parts_group)

        return ret
